### PR TITLE
Remove one host sync from SFT chunked_nll loss

### PR DESCRIPTION
**Pull request by qgallouedec**
`loss_type="chunked_nll"` is the default since #5997 and already wins on VRAM, but it does not win on speed, which is why #6571 reaches for a `compute_token_metrics=False` escape hatch.

This PR removes one of the two device-to-host synchronizations in `_chunked_cross_entropy_loss`, applies the same change to the distillation copy, and reports what the remaining one costs.

## What changed

`_chunked_cross_entropy_loss` forced **two** device-to-host syncs per forward, both on the critical path (measured with `torch.cuda.set_sync_debug_mode("warn")` on an H100):

1. `if n_valid_tensor == 0:`: compared a device tensor in a Python `if`, on every forward, to special-case a fully-masked micro-batch.
2. `for start in range(0, n_padded, chunk_size)`: `n_padded` is a device tensor, so `__index__` blocks.

Sync 1 is now gone. The fully-masked case folds into the general path: `n_padded` is clamped to at least one chunk, so a whole-masked micro-batch yields a zero loss that is still connected to `hidden_states`, `lm_head_weight` and `lm_head_bias` through the chunk matmul, and the mean reduction divides by `n_valid_tensor.clamp(min=1)` so it stays a finite zero. Without that, the loop runs zero chunks, the loss carries no `grad_fn`, and `.backward()` raises.

That case is not rare. Under context parallelism the sequence is split across ranks, so with `completion_only_loss=True` and a long prompt most ranks hold only masked positions. For a 4096-token sequence with a 3900-token prompt, 3 of 4 ranks at `cp_size=4` and 7 of 8 at `cp_size=8` see all -100, under both contiguous and load-balanced sharding. SFT is the trainer that supports CP, which is what the `shift_labels` contract is for. `num_items_in_batch` is deliberately left unclamped: it is a global count, so it stays positive even when a rank is fully masked.

## Why sync 2 stays

Sync 2 buys the masked-chunk skip: valid tokens are packed to the front, and the loop runs only over chunks covering the valid prefix. Deriving the trip count from the static shape instead would remove the sync but pay every masked chunk's `lm_head` matmul. Measured at equal work (Qwen3-0.6B, V=151936, bs=4, seq=2048, chunk=1024, H100):

| variant | full-valid | completion-only (75% masked) |
|---|---|---|
| data-dependent trip count (1 sync) | 491 ms | **288 ms** |
| static trip count (0 syncs) | **479 ms** | 478 ms |

The sync costs ~13 ms/step (2.6%). The skip it enables is worth ~190 ms/step (40%) with `completion_only_loss=True`. Removing it would be a large regression on exactly the config SFT users run, so it stays, and the docstring now says so instead of implying the path is sync-free.

One docstring claim did check out: chunk-count quantization really does bound recompilation. 40 distinct valid-token counts produce 9–10 compilations, not 40. So the remaining cost on XLA is one graph execution per forward, not unbounded retracing.

Removing sync 1 is throughput-neutral on GPU (478.1 vs 478.2 ms), sync 2 immediately drains the queue anyway. Its value is the graph break and one less XLA execution.

## DistillationTrainer

`_chunked_divergence_loss` carries the same duplicated block, same early return and same trip count, so AGENTS.md requires the copies stay aligned and the change is applied there too. DistillationTrainer rejects CP, so it cannot reach the fully-masked-rank case, but the block is kept identical rather than allowed to diverge.

## Ask

@michaelbenayoun the two syncs you suspected are real and this branch removes one of them; the second is a deliberate GPU/XLA trade-off documented above. Could you re-run the exact two-run profile you posted on #6571 against this branch on Trainium?

---

> [!NOTE]
> **Medium Risk**
> Touches the default SFT loss path and distributed training edge cases (fully masked ranks); behavior is intentional but changes when zero valid tokens are handled.
> 
> **Overview**
> Removes a per-forward **device-to-host sync** in the default `chunked_nll` loss by dropping the Python branch on `n_valid_tensor == 0` in `_chunked_cross_entropy_loss` and the matching `_chunked_divergence_loss` in distillation.
> 
> Fully masked micro-batches (common under **context parallelism** with `completion_only_loss`) now always run **at least one** checkpointed chunk: `n_padded` uses `.clamp(min=1)`, and local mean reduction divides by `n_valid_tensor.clamp(min=1)` so the loss stays a finite zero with a valid autograd path for DDP/FSDP. The SFT docstring now states that one sync remains when the chunk loop trip count is read on the host.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b25ce9d56d886e0db72634a3eac79db8fc15a9a8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -226,7 +219,7 @@ def _chunked_cross_entropy_loss(
         entropy_sum = entropy_sum + chunk_entropy
 
     if num_items_in_batch is None:
-        loss = loss / n_valid_tensor
+        loss = loss / n_valid_tensor.clamp(min=1)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:224]: Good catch, fixed in 67d6bfd6: both denominators are clamped now, so a fully masked batch reduces to a finite zero instead of NaN.

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -187,24 +187,17 @@ def _chunked_cross_entropy_loss(
...
     order = valid.to(torch.int8).argsort(descending=True, stable=True)
     hidden = hidden[order]
     labels = labels[order]
 
-    # Process only the whole chunks covering the valid prefix: bounds XLA recompiles and drops fully-masked chunks on GPU.
-    n_padded = (n_valid_tensor / chunk_size).ceil().to(torch.int64) * chunk_size
+    # Process only the whole chunks covering the valid prefix: bounds XLA recompiles and drops fully-masked chunks on
+    # GPU. At least one chunk always runs, so a whole-masked micro-batch still produces a zero loss connected to every
+    # trainable parameter, which `.backward()` and DDP / FSDP gradient sync require.
+    n_padded = (n_valid_tensor / chunk_size).ceil().clamp(min=1).to(torch.int64) * chunk_size
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:201]: Right, propagated in 80af1a4c: same early return removal, chunk clamp and reduction in _chunked_divergence_loss, outputs unchanged.

File: trl/trainer/sft_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:224]: actually, not taking this one. num_items_in_batch sums non-masked labels over the accumulation window, and TRL cannot produce a window where that is 0. EOS is appended to every completion, so even completion="" leaves one valid label, and examples left fully masked by truncation are filtered out during dataset preparation. I ran the completion-only plus truncation case you describe: the row survives as [-100 x7, 151645] and num_items_in_batch is 1, loss finite. Clamping would be defensive code for a state that does not exist, so c2cdf1a6 removes the guard. If it ever did occur, the loop runs zero chunks and backward raises, which is loud rather than silent.

**Comment by michaelbenayoun** [comment]
Thanks @qgallouedec. Profiling it on Neuron.

**Comment by michaelbenayoun** [comment]
After profiling it seems a bit faster.

image

Nothing crazy but it was to be expected. At least it runs. I will validate with a full training.

Also a bunch of ops in the `_chunk` function seem to be running on CPU. Still not clear to me if it is because the ops are not supported on Neuron yet or something else. Investigating.

**Comment by qgallouedec** [comment]
Merging now, the rest is follow-up work.

The remaining sync stays: the data-dependent trip count buys the masked-chunk skip, worth ~190 ms/step with completion_only_loss=True against ~13 ms/step for the sync itself.

@michaelbenayoun thks for profiling. Keep me posted on the full training run and on the ops falling back to CPU in `_chunk`. If something needs to change on the TRL side I'll take it as a follow-up.

Next on my end, as discussed offline with you: the per-step metric collectives around the loss. Accumulate per rank during the step and gather once in log(). Separate PR coming
